### PR TITLE
Only run workflows if they might be able to work

### LINKED ISSUE #63163: CI fails in forks because it uses secrets unconditionally
**What happened**:

Lots of workflows fail when triggered in forks

**What you expected to happen**:

* workflows should declare permissions they need
* workflows should verify they have any secrets they need before trying to do work that requires them

**How to reproduce it (as minimally and precisely as possible)**:

1. create an organization
2. set the organization permissions to readonly instead of read write https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#configuring-the-default-github_token-permissions
3. create a fork of the grafana repository
4. go to actions and enable actions (there are ways that they'll be automatically triggered)
5. you can just wait for a scheduled task to run, or you can trigger various events (e.g. creating an issue in your fork w/ a label, or creating a PR in your fork)

**Anything else we need to know?**:

**Environment**:
- Grafana version: main
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by jsoref**
**What is this feature?**

Add permissions to workflows that need permissions and check for secrets in workflows that use secrets

**Why do we need this feature?**

Otherwise forks get lots of failures

**Who is this feature for?**

Contributors who aren't members of the grafana organization and thus need to use forks where github actions may run, and especially those who run those forks in organizations that are moderately concerned about security (and thus have restricted the github_token's default permissions)

**Which issue(s) does this PR fix?**:

Fixes #63163

**Special notes for your reviewer**:

You'll want to test this in your own organization or repository. I ran through each workflow that I changed and made sure it ran at least once without secrets and then I ran at least one with some mock secrets to confirm that the run case works.

### CODE REVIEW

**Review by DanCech — COMMENTED** [review]
Looks reasonable to me, thanks!

For other reviewers:
- `GH_TOKEN` is the [preferred env var](https://cli.github.com/manual/gh_help_environment) for the `gh` cli tool

File: .github/workflows/detect-breaking-changes-report.yml
Comment (leventebalogh) [inline .github/workflows/detect-breaking-changes-report.yml:19]: Hmm, why do we actually need these explicit permissions? (I mean in what scenarios were the current permissions of the workflow not sufficient?)

File: .github/workflows/detect-breaking-changes-report.yml
Comment (leventebalogh) [inline .github/workflows/detect-breaking-changes-report.yml:126]: I am not really sure about these checks to be honest. 

If the reporting to Slack or to the Github PR would be optional, then I think this would make a lot of sense, however we would like to be notified about breaking changes, and if the reporting breaks due to a missing permission / github token, then we would like to know about it. If we introduce this change then it will be harder to spot these issues, as the workflow steps will be just skipped instead of failing.

File: .github/workflows/detect-breaking-changes-report.yml
Comment (jsoref) [reply .github/workflows/detect-breaking-changes-report.yml:19]: Organizations with restrictive default permissions as noted in the issue.

Here's an example:
https://github.com/check-spelling/grafana/actions/runs/4130829125/jobs/7137892469#step:1:17

File: .github/workflows/detect-breaking-changes-report.yml
Comment (jsoref) [reply .github/workflows/detect-breaking-changes-report.yml:126]: If you want it to fail just for the grafana instance, I can add an additional condition. But as long as you *have* a credentials set, you'll still get a failure, it's only if you don't have one set at all that you won't.

I'd find it hard to believe that you wouldn't notice getting zero notifications to Slack if you were still using Slack/notifications. (We use Slack heavily.)

File: .github/workflows/detect-breaking-changes-report.yml
Comment (leventebalogh) [reply .github/workflows/detect-breaking-changes-report.yml:19]: Fair enough 👍

File: .github/workflows/detect-breaking-changes-report.yml
Comment (leventebalogh) [inline .github/workflows/detect-breaking-changes-report.yml:124]: This is going to break the workflow (as env is now defined twice, check the error message [here](https://github.com/check-spelling/grafana/actions/runs/4130891051)). 

Just merge the two `env` declarations, (this and the one at line `137`).

File: .github/workflows/detect-breaking-changes-report.yml
Comment (leventebalogh) [reply .github/workflows/detect-breaking-changes-report.yml:126]: We are trying to build tools that we don't need to check constantly to see if they are still working ("there were no notifications since 2 days, can it be that it's broken?"). This workflow is reporting breaking changes and it is not always happening that often that we would immediately see if it goes silent and misses a few PRs. Even if we would notice it after some time, missing a "heads up" for a possible breaking change can cause us head aches.

I have a feeling this workflow does only make sense in the main grafana repo, and don't have any added value for forks, so we could probably introduce a condition (e.g. github.repository == "grafana/grafana") to stop failing on forks?

Comment (jsoref) [reply .github/workflows/detect-breaking-changes-report.yml:126]: You could certainly do that.

Having just had someone yell at me in another repository for making a PR with "breaking changes" but who couldn't be bothered to explain what the breaking changes were, I'd much rather the CI tell me what the breaking changes are in my fork before I make a PR to upstream.

As such, my personal preference would be to just include that into the HAS_SECRETS condition so that it could be used to send a message to my slack instance if I so chose.

That said, my goal with this PR is to not drown in :x: (see https://github.com/jsoref/grafana/actions) in my forks. I'll do whatever makes this repository happy.

**Review by leventebalogh — COMMENTED** [review]
Although looks good to me for the first glance, I think it would be the best if the code-owners of the affected workflows could have a look as well. Listing below:

- [x] @grafana/grafana-release-eng
    - [x] `backport.yml`
    - [x] `bump-version.yml`
    - [x] `close-milestone.yml`
    - [x] `update-changelog.yml`

- [ ] @torkelo
    - [ ] `commands.yml`
    - [ ] `github-release.yml`
    - [ ] `metrics-collector.yml`

- [x] @grafana/plugins-platform-frontend
    - [x] `detect-breaking-changes-report.yml`

- [ ] @tolzhabayev
    - [ ] `pr-commands-closed.yml`

- [x] @armandgrillet
    - [x] `issue-labeled.yml`

- [ ] @meanmina
    - [ ] `epic-add-to-platform-ux-parent-project.yml`

- [ ] @grafana/docs-grafana
    - [x] `publish-technical-documentation-next.yml`
    - [x] `publish-technical-documentation-release.yml`

- [ ] @marefr
    - [ ] `milestone.yml`
    - [ ] `pr-checks.yml`
    - [ ] `pr-commands.yml`

- [ ] @grafana/user-essentials
    - [ ] `remove-milestone.yml`
    - [ ] `stale.yml`

- [x] @DanCech
    - [x] `codeql-analysis.yml`
    - [x] `pr-codeql-analysis-go.yml`
    - [x] `pr-codeql-analysis-javascript.yml`
    - [x] `pr-codeql-analysis-python.yml`

**Review by torkelo — COMMENTED** [review]
Not sure about this, some of these workflows are not run for PRs / external contributors, so not sure these checks are needed?

**Comment by torkelo** [comment]
@jsoref is there a clean way to limit them to this repo/fork then?

File: .github/workflows/detect-breaking-changes-report.yml
Comment (jsoref) [inline .github/workflows/detect-breaking-changes-report.yml:137]: @torkelo here's an example of using an `if` to check for a specific repository.

If you don't want things to run on other repositories, you'd just use more of this logic and less of the rest.

**Comment by tolzhabayev** [comment]
It feels harder to read, so maybe we can either:
- Add `if: github.repository == 'grafana/grafana'` to the actions that should not run on forks instead of has_secrets checks.
- Suggest to forks to disable github actions all together - e.g. https://github.com/orgs/community/discussions/26704#discussioncomment-3252979

**Comment by jsoref** [comment]
Each repository/ organization has different policies, but generally most get annoyed when contributions would fail most CI. Thus it's really helpful/important to be able to run some of it before making a pull request.

It's also frustrating from my perspective as someone who tries to make drive by contributions, it's often quite expensive to set up local build environments for each specific repository, whereas GitHub workflows maintained by upstream tend to work fairly reliably (and when they don't, I'll work with upstream to improve them, as here). Asking people to actively turn off actions really isn't a reasonable approach. 

If you have specific workflows that should only be run in this repository, just identify them and I'll adjust this PR sometime this week.

**Comment by kminehart** [comment]
Most of our GitHub actions don't do any code validation or testing and are more for process management, like ensuring that your pull request is titled accurately and has a milestone, for example. We use drone (drone.grafana.net) to run our tests and it does actually run builds for forks when a pull request is opened.

> If you have specific workflows that should only be run in this repository, just identify them and I'll adjust this PR sometime this week.

Here are the ones that I can easily identify.

```
backport.yml
bump-version.yml
close-milestone.yml
update-changelog.yml
```

Are all for managing our branching strategy / versioning / changelog / documentation, so I don't think these have any value running on forks.

```
publish-technical-documentation-next.yml
publish-technical-documentation-release.yml
```

Similarly, these will not run without write access to our a private repo of ours, so I don't see these ever being useful to execute in a fork.

As for the rest I don't personally know enough about them.

File: .github/workflows/publish-technical-documentation-next.yml
Comment (kminehart) [inline .github/workflows/publish-technical-documentation-next.yml:11]: I'm not super familiar with GitHub actions; is changing this from 'sync' to 'main' essentially a no-op?

Comment (jsoref) [reply .github/workflows/publish-technical-documentation-next.yml:11]: Job ids have no particular value unless you're referencing them for things like `needs:` / `needs.` or `jobs.{...}.outputs` or similar.

I didn't mean to change this, I'll undo this change.

**Comment by jsoref** [comment]
Sigh

**Comment by jsoref** [comment]
@kminehart I've rebased this (again)

**Comment by armandgrillet** [comment]
The changes on the workflow I maintain make sense. 

This PR is nearly 4 months old, due to Grafana 10.0 being cut I think we should merge and fix on the fly if some automation breaks instead of waiting for everyone to review due to the number of contributors on this repository. @torkelo could you take another look at this PR and give us your opinion? @jsoref explained why such workflows are frustrating for fork maintainers in comments.

**Comment by jsoref** [comment]
Sigh. I'll try to look tonight

**Comment by jsoref** [comment]
@armandgrillet ...

File: .github/workflows/codeql-analysis.yml
Comment (armandgrillet) [inline .github/workflows/codeql-analysis.yml:22]: This allows an action to update the status of a code scanning alert, this makes sense.

**Review by armandgrillet — APPROVED** [review]

**Comment by armandgrillet** [comment]
Due to main now targeting 10.2 which will not be released anytime soon, reviewing this and not finding anything that does not look like an improvement making the life of external contributors easier, and the lack of other reviews since my last comment: I am merging this PR now.

**Pull request merged**
